### DISCUSSION TITLE
shelly plus addon analog input

### DISCUSSION THREAD

**Discussion by ChrisROL** [arendst/Tasmota#21248] [Q&A]
Hi,

is there a possibility to use the shelly plus addon analog input on gpio 3 with an ntc resistor?
As far as i understood , the input works with an sawtooth comperator and the signal on gpio3 should be simular to an pwm signal.
Maybe it can be done in berry, but i have not worked with berry till now.

Regards Chris

**Comment by ChrisROL** [comment]
Hi there,

has really nobody any idea how to make the "analog" input at a shelly plus addon useable?

**Reply by HofeBY** [reply]
Why not using as a digital GPIO Port with up to several DS18B20 ?
https://www.amazon.de/s?k=DS18B20&__mk_de_DE=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=NX8LU4MV8MF6&sprefix=ds18b20%2Caps%2C221&ref=nb_sb_noss_1

**Reply by ChrisROL** [reply]
Hi,

the reason is, i want to replace an old thermostat and the ntc is already installed and i'm not shure if it is possible to replace it.
And i think, there is more than my case where the analog input at the addon maybe helpfull. Or is there any PWM input with TASMOTA?

**Reply by HofeBY** [reply]
Which CPU you are using ?
PWM (Pulse Width Modulation) are output from CPU side.
Is your NTC are a type of 10kOhm with a ß=4000K ? Used for Shelly Plus Addon Modul.

**Reply by sfromis** [reply]
I saw someone suggest this:
> The board is generating a sawtooth input to one pin of the comparator, with the analog input on the other, giving a digital output with duty cycle representing the analog input.

AFAIK, Tasmota has no support for measuring duty cycle (ratio of time high/low), the closest feature is counters providing a count of pulses, but that does not make sense with what appears to be a fixed frequency.

Measuring the duty cycle would take interrupts on transitions to both high and low, and time calculations.

In principle, it just might be feasible in Berry, but I'd expect the sawtooth frequency to be too high for that.

**Reply by ChrisROL** [reply]
This is exactly why I'm asking about a PWM input for Tasmota. I don't know how Berry is implemented in Tasmota and whether interrupts are supported. I had solved most of my previous problems using GEMU scripting

**Reply by sfromis** [reply]
And I answered about PWM input. Not supported. Interrupts are only workable if you create a native driver in C++, along the lines of the existing driver supporting pulse counting via interrupts.

ESP32 also has a pulse counter feature (PCNT) for counting pulses, but AFAIK this is not used by Tasmota, and while I'm not sure of any details, a cursory look does not look like this is supporting duty cycle.

**Reply by barbudor** [reply]
@ChrisROL an NTC will not generate an PWM signal. You would need additional circuitry for that.
The 'add-on' input are not straight connected to an ESP GPIO so you can't use an analog input

So instead of making it more complicated than it is, use what is supported by the hardware and the software.
Otherwise it's that you choice of hardware (the Shelly) is not appropriate to your project

**Reply by sfromis** [reply]
@barbudor What is suggested by what I saw is that the plus addon indeed has an analog input, which is not connected to an analog pin on the ESP32. Instead the addon has additional circuitry generating a sawtooth pulse, as one side input to a comparator, thus generating a PWM signal with the duty cycle matching the analog voltage.

Cool idea, I suppose, but it is only workable if the software on the ESP32 is able to detect the duty cycle. Which should not be a "big problem", if *someone* decided to code a driver to handle that, maybe on top on the counter driver which can be used to determine frequency.

Yeah, Shelly does seem to like making things more complicated, instead of just providing simple gpio connections.....

**Reply by ChrisROL** [reply]
I think shelly goes this way, because there power supply for the ESP is not isolating the 3,3V from the 230V.
I guess the duty cycle measuring shoul be possible, the original shelly firmware does support it.
Unfortunately, with my very limited programming skills, I am unable to adapt the counter driver

**Comment by Jason2866** [comment]
The pragmatic way, replace the Shelly with an esp32 and use the adc ntc feature implemented in Tasmota.

**Reply by ChrisROL** [reply]
Sound very easy, but there is a reason why i choiced the shelly plus with addon.
In a very simular room with nearly the same conditions i did it with a LUANI HVIO Modul and a printed case.
But in this room the heater mat needs 230V/7A and the controller should be placed in a german flush-mounted box
If you know a device, better matching to my requirement  than the shelly, please tell me.  I'm not fixed to shelly but the NTC and the heating mat are not really changeable.

**Comment by sohailb108** [comment]
Another - may be overkill - solution would be to simulate a DS18B20 using a micro-controller. 
There are good comments and links here:
https://github.com/arendst/Tasmota/discussions/17133

**Reply by ChrisROL** [reply]
Hi,
it's me again. Has somebody a suggestion for a different hardware for my floor heating?
I have a NTC for the floor temperature, a DHT22 for the room temperature (could be changed) and need a relais for 230V min. 7A .
Or is somebody able to do the programming of a additional driver? If necessary, i would also provide my hardware components.
